### PR TITLE
[Bug] Add SharedCycleIteratorState

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

The following script will create an infinite loop on master. 

```py
from torch.utils.data.sampler import SequentialSampler
from pytorch_lightning.trainer.supporters import CombinedLoader
from torch.utils.data import IterableDataset, DataLoader, Dataset

class MyIterableDataset(IterableDataset):

    def __init__(self, size: int = 10):
        self.size = size

    def __iter__(self):
        self.sampler = SequentialSampler(range(self.size))
        self.iter_sampler = iter(self.sampler)
        return self

    def __next__(self):
        return next(self.iter_sampler)

class MyMapDataset(Dataset):

    def __init__(self, size: int = 10):
        self.size = size

    def __getitem__(self, index):
        return index

    def __len__(self):
        return self.size

loaders = [DataLoader(MyIterableDataset(6)), DataLoader(MyMapDataset(5))]
dataloader = CombinedLoader(loaders, mode="max_size_cycle")
for _ in dataloader:
    print("i")
```

This PR adds a `SharedCycleIteratorState` to synchornize the CycleIterator and stop on max_length.

Fixes #\<issue_number>

### Does your PR introduce any breaking changes? If yes, please list them.

No.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/supporters.py [hunk 1]
```diff
@@ -205,22 +227,30 @@ def __next__(self) -> Any:
...
-
         Returns:
             Any: the resulting batch
-
         Raises:
             StopIteration: if more then :attr:`length` batches have been returned
         """
         # Note: if self.length is `inf`, then the iterator will never stop
-        if self.counter >= self.__len__():
+        if self.counter >= self.__len__() or (self.state is not None and self.state.done):
```
Comment (justusschock) [inline pytorch_lightning/trainer/supporters.py:236]:
```suggestion
        if self.counter >= self.__len__() or self.state.done:
```

shouldn't we simply always create a state like this?

File: pytorch_lightning/trainer/supporters.py [hunk 2]
```diff
@@ -205,22 +227,30 @@ def __next__(self) -> Any:
...
             raise StopIteration
 
         try:
             return next(self._loader_iter)
 
         except StopIteration:
+
+            if self.state is not None:
```
Comment (justusschock) [inline pytorch_lightning/trainer/supporters.py:244]: See above. Let's always create a state for consistency

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/supporters.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py]: Done !

File: pytorch_lightning/trainer/supporters.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/supporters.py]: Done !

**Review by ananyahjha93 — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/trainer/supporters.py [hunk 3]
```diff
@@ -170,12 +171,35 @@ def to_disk(self) -> None:
                 torch.save(outputs, fp)
 
 
+@dataclass
+class SharedCycleIteratorState:
+
+    mode: str = "max_size_cycle"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/supporters.py:177]:
```suggestion
class SharedCycleIteratorState:
    """A state shared between all CylceIterators in a CombinedLoader. 
    
    With a shared state, the iterators can decide to terminate based on the state of all others.
    If the mode is *max_size_cycle*, all iterators need to have finished before the combined loading is considered
    finished, and otherwise any iterator finishing early will lead to all iterators ending early.
    """
    mode: str = "max_size_cycle"
```

Comment (awaelchli) [reply pytorch_lightning/trainer/supporters.py:177]: What do you think of these docs? Did I get it right?

**Review by awaelchli — APPROVED** [review]
